### ISSUE TITLE
Add pictorial route shields

### ISSUE THREAD

**Issue by 1ec5** [gravitystorm/openstreetmap-carto#508]
In many parts of the world, the cartographic convention is to mark routes with graphical shields resembling road signage. This renderer marks routes with plain text badges that are suitable for countries that identify routes with consistent alphabetic prefixes, such as the U.K.’s M/A/B scheme. However, in countries that use graphical shields, this style makes OSM look foreign and much less impressive to potential contributors.

In the U.S., this renderer’s plain text badges often lead to disagreements and edit wars about the proper format of `ref` tags on ways. Some would prefer the `ref` tag to reflect real-world written usage (e.g. “I-10” for an Interstate route or “SR 123” for a state route), while others have pushed for a uniform, machine-readable syntax with a unique alphabetic prefix for each route network (“I 10” and “CA 123”). The U.S. is currently a patchwork of these schemes.

Meanwhile, we already have very good penetration of `type=route` relations, which solve this problem with a separate, machine-readable `network` tag. But we still squabble over the `ref` tag on ways because route relations are ignored by mainstream renderers.
#495 requested arbitrary route shields based on URLs in the `symbol` tag, but that suggestion should scare anyone who wants to deploy a renderer based on live OSM data. Instead, we should maintain a large set of shield images in the symbols directory and select from these images using the `network` tag on route relations.

@asciiphil has built a [fully-functional prototype](http://elrond.aperiodic.net/shields/) ([source](https://code.launchpad.net/%7Easciiphil/osm-shields/trunk)) based on this idea. This renderer currently supports [a variety](http://elrond.aperiodic.net/shields/supported.html) of national, state, and some county and township routes in the U.S.; provincial routes in Canada; federal routes in Mexico; and some Australian routes. It also handles concurrencies and banners, which are difficult to express with `ref` tags on ways. It could easily be extended to handle country-specific shields in Europe (such as the blue hexagons in Germany), provided that route relations are used. The shield renderer is popular among members of the talk-us mailing list, but it needs to be ported over to the new Carto stylesheets.

![Mapnik](http://tile.openstreetmap.org/14/4334/6104.png) ![Shields](http://web.archive.org/web/20180102090956if_/http://tile.openstreetmap.us/osmus_shields/14/4334/6104.png)

![Mapnik](http://b.tile.openstreetmap.org/14/4347/6254.png) ![Shields](http://tile.openstreetmap.us/osmus_shields/14/4347/6254.png)

![Mapnik](http://b.tile.openstreetmap.org/16/17094/24880.png) ![Shields](http://b.tile.openstreetmap.us/osmus_shields/16/17094/24880.png)

The shield renderer stores a series of public domain “template” SVG images which each have SVG text placeholders for route numbers. A “sequences” folder contains files that list ranges of valid route numbers for each route network. A preprocessing step assembles SVG images for each valid route number for each network. The Mapnik stylesheets look for these images.

For countries that rely on graphical shields, rendering them accurately is a functional requirement, not an aesthetic one. Please consider implementing functionality similar to what the shield renderer provides (if not necessarily the same pipeline). Thank you!

**Comment by kocio-pl** [comment]
Do we really want to do this? I'm skeptical, because we don't do such localization for any other feature and the roads code is already big and complex enough.

**Comment by pnorman** [comment]
We want this, but an effective implementation depends on #596.

**Comment by 1ec5** [comment]
> Given how @asciiphil's proof of concept works, I doubt we'll end up reusing any code from it.

@kennykb started [osm-shields](https://github.com/kennykb/osm-shields/) as a successor to @asciiphil’s prototype that is intended to be more usable in a standard Mapnik-based renderer. I haven’t dug into the details, but hopefully it offers a way forward on this issue.

**Comment by kennykb** [comment]
@1ec5  - Maybe. I wanted pictorial route markers on my own maps, and it was such a long road figuring out how to make Mapnik's GroupSymbolizer do my bidding that I thought it was best to start a project so that at least the workflow was documented for others. (And so that I wouldn't forget how it worked!) If it turns out to be useful for OSM Carto, that would be a plus, but I'm not actively lobbying for that at present. I haven't the time and energy for that - it's a very high bar to clear.

**Comment by nvkelso** [comment]
We open sourced 100s of road shield artwork as Mapzen was shutting down for all USA state shields and ~50 countries around the world (federal down to state and sometimes county). 

See the [blog post](https://medium.com/postzen/long-term-support-for-mapzen-maps-vector-terrain-tiles-a45e108943b4) (scroll down to the bottom) and artwork [repo](https://github.com/tangrams/icons) (**Bubble Wrap** is farthest along, and the map preview shows a [demo](https://tangrams.github.io/bubble-wrap/#11/37.7926/-122.4001) using Tangram). The artwork is vector in Adobe Illustrator format and exported in raster. Can easily be converted to other formats.

**Comment by pnorman** [comment]
We can't use GroupSymbolizer. It isn't supported by CartoCSS. I don't think shield pictures are a serious blocker for this feature, given the other difficult problems faced.

**Comment by 1ec5** [comment]
> We can't use GroupSymbolizer. It isn't supported by CartoCSS.

Ah, good point. mapbox/carto#349 was proposed a long time ago to address that, and seeing that reminds me that I had proposed an alternative syntax (1ec5/carto@1c9ff62ceb06e8a8b5abf5762f9d296472541602) but never followed through with a PR. 🙈

**Comment by jeisenbe** [comment]
@1ec5 were you still considering submitting a new PR to carto so that we can use GroupSymbolizer in CartoCSS?

**Comment by JimFlasch** [comment]
How close are we to having proper highway shields on OpenStreetMap by now?

**Comment by matkoniecz** [comment]
In this style? Quite far away. Unless at least "We can't use GroupSymbolizer. It isn't supported by CartoCSS." was fixed or some alternative was found.

Though there was work in https://github.com/ZeLonewolf/openstreetmap-americana using a different technical stack, one of main planned features are US-centric highway shields.

**Comment by ZeLonewolf** [comment]
> How close are we to having proper highway shields on OpenStreetMap by now?

As @matkoniecz notes, we expect to have working highway shields in openstreetmap-americana, which is a vector style, shortly after OpenMapTiles completes its 3.13 release, which will include support for concurrent highway routes.  This was completed in openmaptiles/openmaptiles#1152.  However, they will not yet be aligned properly until we can get a fix for maplibre/maplibre-gl-js#188.  So, this is actively being worked in other technical stacks, but I'm not aware of anyone actively working on highway shields for raster stacks like the one used by openstreetmap-carto.

**Comment by attilakundev** [comment]
> Thank you for this well-documented request.
> 
> How do paper maps handle this? I suppose US paper maps show the shields? I can't remember ever having seen shields rendered on map in continental Europe. What about the UK?


https://map.hu/wp-content/uploads/2020/11/2271_hu_450k_2013_puha_balaton.jpg
https://mindentudasboltja.hu/Magyarorszag-autoterkep
https://www.avenzamaps.com/maps/897982?utm_source=affiliate&utm_medium=affiliate_link&utm_campaign=szarvas.andras%40map.hu&utm_term=2201846899

 @matthijsmelissen Sorry, but your your statement is totally wrong, because hungarian road atlases DO display road numbers and even the E-roads. (And Hungary is in Europe)

**Comment by pnorman** [comment]
#4431 will unblock #596 which unblocks this.

**Comment by ZeLonewolf** [comment]
The OpenStreetMap Americana project is a CC0 renderer that is working towards a comprehensive rendering of highway shields.  As a CC0 project, this work is license-compatible with this project and osm-carto contributors should feel free to re-use any of this work that may be applicable.

Shield artwork can be found here:
https://github.com/ZeLonewolf/openstreetmap-americana/tree/main/icons

Network and shield definitions can be found here:
https://github.com/ZeLonewolf/openstreetmap-americana/blob/main/src/js/shield_defs.js

**Comment by imagico** [comment]
In addition to #596 - what would be really helpful for any map design project looking into doing something like suggested here: A library containing all the pertinent graphical and classification information on pictorial road shields in a generic and non-obscure form (In other words: In a form that semantically well defined, that is strait away to write a converter for to translate it into whatever other form you need it in and that is not tied to a specific use case - like a certain design of map, or even cartographic applications in general).  This would include the shield geometries and colors of course, but also information on labeling rules and other dynamic properties.

That would be a project of significant proportions of course.

**Comment by michaelblyons** [comment]
> A library containing all the pertinent graphical and classification information on pictorial road shields in a generic and non-obscure form

ICYMI, Americana publishes the shield renderer [as an NPM package](https://www.npmjs.com/package/@americana/maplibre-shield-generator). Perhaps a helpful milestone.
